### PR TITLE
PRIME-2499 Device Provider Enrolment

### PR DESCRIPTION
**Pull request by bergomi02**

### CODE REVIEW

**Comment by neophyte57** [comment]
@bergomi02 In the PR environment, for this user, if I go Back from the College Licence page, and then go Save and Continue, I get a client-side error

image

**Comment by bergomi02** [comment]
> @bergomi02 In the PR environment, for this user, if I go Back from the College Licence page, and then go Save and Continue, I get a client-side error
> 
> image

Yes, I noticed that too. but I can't find here the problem is.

**Comment by neophyte57** [comment]
Doing regression testing, enrolling physician, I get this on Save and Continue

image

File: prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html [hunk 1]
```diff
@@ -12,31 +12,86 @@
     <section *ngIf="isDeviceProvider"
              class="mb-3">
 
+      <app-page-subheader>
+        <ng-container appPageSubheaderTitle>What is your role?</ng-container>
+        <ng-container appPageSubheaderSummary>
+          Please select “none” to continue if your role is not listed.
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html:18]: Should capitalize, “none” to “None”

File: prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html [hunk 2]
```diff
@@ -12,31 +12,86 @@
...
+      <div class="row">
+        <div class="col col-sm-10 py-3">
+          <mat-form-field class="w-100">
+            <mat-label>Device Provider ID</mat-label>
+            <input matInput
+                   maxlength="8"
+                   placeholder=""
+                   formControlName="deviceProviderId">
+            <mat-error *ngIf="formState.deviceProviderId.hasError('deviceProviderId')">
+              Must be in format P1-90XXX (The last three digits are numeric)
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html:57]: Wondering whether the hint should be "(last three characters are numbers)"?

Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html]: Text updated as suggested

File: prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html (hunk 1 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.html]: fixed

**Comment by bergomi02** [comment]
> Doing regression testing, enrolling physician, I get this on Save and Continue
> 
> image

it is fixed

**Comment by bergomi02** [comment]
> @bergomi02 In the PR environment, for this user, if I go Back from the College Licence page, and then go Save and Continue, I get a client-side error
> 
> image

this issue should be fixed.

File: prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.ts [hunk 3]
```diff
@@ -133,16 +166,49 @@ export class RegulatoryComponent extends BaseEnrolmentProfilePage implements OnI
...
+  public onSubmit() {
+    if (this.formUtilsService.checkValidity(this.form)) {
+
+      if (this.isDeviceProvider && this.formState.deviceProviderId.value) {
+        let siteName = this.deviceProviderSite ?
+          this.deviceProviderSite.siteName : "Site not found.";
+        let siteAddress = this.deviceProviderSite ?
+          `${this.deviceProviderSite.siteAddress}, ${this.deviceProviderSite.city}, ${this.deviceProviderSite.prov}` : "";
+        let message = this.deviceProviderSite ?
+          "Is this the correct location where you are the device provider?" : "If you are not sure the Device Provider ID, you might continue."
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.ts:182]: Would read better if "If you are not sure of the Device Provider ID, you may continue."

File: prime-dotnet-webapi/Models/SelfDeclarationVersion.cs [hunk 4]
```diff
@@ -17,5 +17,7 @@ public class SelfDeclarationVersion : BaseAuditable
         public string Text { get; set; }
 
         public DateTimeOffset EffectiveDate { get; set; }
+
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/SelfDeclarationVersion.cs:20]: Needs a comment ... "A comma-separated list of all care setting codes for which applicable"?

File: prime-dotnet-webapi/Models/DeviceProviderRole.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Prime.Models
+{
+    /// <summary>
+    /// used for device provider
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/DeviceProviderRole.cs:7]: "Used for Device Provider Individual Enrollment" instead?

File: prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.ts (hunk 3 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/regulatory/regulatory.component.ts]: Good call, updated.

File: prime-dotnet-webapi/Models/SelfDeclarationVersion.cs (hunk 4 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Models/SelfDeclarationVersion.cs:20]: Good call, updated.

File: prime-dotnet-webapi/Models/Enrollee.cs [hunk 6]
```diff
@@ -67,6 +67,8 @@ public Enrollee()
 
         public ICollection<EnrolleeCareSetting> EnrolleeCareSettings { get; set; }
 
+        public ICollection<EnrolleeDeviceProvider> EnrolleeDeviceProviders { get; set; }
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/Enrollee.cs:70]: Why is this a 1-to-many relationship?

File: prime-dotnet-webapi/Models/EnrolleeDeviceProvider.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        [JsonIgnore]
+        public int EnrolleeId { get; set; }
+
+        public string DeviceProviderId { get; set; }
+
+        public string CertificationNumber { get; set; }
+
+        [JsonIgnore]
+        public Enrollee Enrollee { get; set; }
+
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/EnrolleeDeviceProvider.cs:22]: Can you move this below `EnrolleeId`?

Comment (bergomi02) [reply prime-dotnet-webapi/Models/EnrolleeDeviceProvider.cs]: fixed

File: prime-dotnet-webapi/Models/Enrollee.cs (hunk 6 above)
Comment (neophyte57) [reply prime-dotnet-webapi/Models/Enrollee.cs:70]: Can you add a comment with some specifics?

Comment (bergomi02) [reply prime-dotnet-webapi/Models/Enrollee.cs:71]: added comment

**Comment by neophyte57** [comment]
When attempting to Share GPID

image

**Comment by neophyte57** [comment]
TOA formatting problems  😑

image

**Comment by neophyte57** [comment]
Problems with display of Provisioner Link web page

image

**Review by neophyte57 — APPROVED** [review]
Looking good!

**Pull request merged**